### PR TITLE
Implement ProductQueryBuilder class to handle product queries

### LINKED ISSUE #52414: [Enhancement]: Implement ProductQueryBuilder class to handle product queries
The current `ProductCollection` class in the WooCommerce Blocks plugin is handling multiple responsibilities, including building and modifying product queries. This makes the class large and difficult to maintain.

To improve code maintainability and adhere to best practices, let's refactor the query-building logic into a separate `ProductQueryBuilder` class. This new class will be solely responsible for constructing and manipulating product queries based on various parameters and filters.

By isolating the query logic, we enhance readability and make future enhancements or optimizations to the querying process more manageable. This change will also facilitate unit testing by allowing us to test query logic independently from other functionalities.

#### What needs to be done

- Extract all methods related to query construction and modification from `ProductCollection` into a new `ProductQueryBuilder` class.
- Ensure `ProductCollection` utilizes `ProductQueryBuilder` for all query-related operations.
- Update relevant unit tests and documentation to reflect the new class structure.
- Verify that all existing functionalities continue to work as expected after the refactor.

### PR DESCRIPTION
**Pull request by octaedro**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR refactors the class Product Collection in the following way:
- All the classes related to Product Collection now live under the folder `ProductCollection.`
- The class `ProductCollection` was renamed to `ProductCollectionController`.
- Some methods were moved from `ProductCollection` to `ProductQueryBuilder`.

This change is part of a bigger change described in [this epic](https://github.com/woocommerce/woocommerce/issues/52418).

Don't be afraid about the lines' number; this PR essentially moves code around.

Closes #52414.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout this branch 
2. Ensure `ProductCollectionController` (ex `ProductCollection`) utilizes `ProductQueryBuilder` for all query-related operations.
3. Verify that all existing functionalities continue to work as expected after the refactor.
4. Test all the Product Collection types.
5. Also test the [No results](https://github.com/woocommerce/woocommerce-blocks/pull/11783/) option.
6. Verify that all the tests pass correctly.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductQueryBuilder.php
Comment (octaedro) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductQueryBuilder.php:55]: We will need this setter to set the `collection_handler_store` at `ProductCollectionController` level. This will change with the introduction of the [class CollectionHandlerRegistry](https://github.com/woocommerce/woocommerce/issues/52417).

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php
Comment (octaedro) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php:835]: We want to set the `collection_handler_store` in the `ProductQueryBuilder` class too.

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductQueryBuilder.php
Comment (octaedro) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductQueryBuilder.php]: We will need this setter to set the `collection_handler_store` at `ProductCollectionController` level. This will change with the introduction of the [class CollectionHandlerRegistry](https://github.com/woocommerce/woocommerce/issues/52417).

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php
Comment (octaedro) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php]: We want to set the `collection_handler_store` in the `ProductQueryBuilder` class too.

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php
Comment (imanish003) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php]: Looks like this is not used anywhere. I think we can remove it.

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php
Comment (imanish003) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php]: Looks like this is not used anywhere. I think we can remove it.

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php
Comment (imanish003) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php]: Looks like we can remove `$render_state` as it's not used anywhere.

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php
Comment (imanish003) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php]: Nitpick: This comment looks outdated.

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductQueryBuilder.php
Comment (imanish003) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductQueryBuilder.php]: I am seeing a lint error here. Are you able to see it? Or it's just me 😅
image

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php
Comment (imanish003) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php]: image

**Review by imanish003 — CHANGES_REQUESTED** [review]
Hey @octaedro,

I quickly went through the code and left some comments. The main goal of splitting the ProductCollection class into smaller files and moving query logic into ProductQueryBuilder looks good. 

I suggest we keep this PR focused solely on the code reorganization (moving code around and splitting the big file) without including any small refactoring changes. This will make the PR easier to review. Once merged, we can iterate on improving each file in follow-up PRs.

Once you address the comments, I can do a final functional testing. If everything works as expected, we can merge this and then work on further improvements.

What do you think about this approach?

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php
Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php]: Fixed in commit [23a56a9](https://github.com/woocommerce/woocommerce/pull/52429/commits/23a56a9db7edd5ea1071b7fabdf27c44c83a7501)

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php
Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php]: Fixed in commit [23a56a9](https://github.com/woocommerce/woocommerce/pull/52429/commits/23a56a9db7edd5ea1071b7fabdf27c44c83a7501)!

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php
Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php]: Fixed in commit [23a56a9](https://github.com/woocommerce/woocommerce/pull/52429/commits/23a56a9db7edd5ea1071b7fabdf27c44c83a7501)!

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php
Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductCollectionController.php]: Fixed in commit [23a56a9](https://github.com/woocommerce/woocommerce/pull/52429/commits/23a56a9db7edd5ea1071b7fabdf27c44c83a7501)!

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductQueryBuilder.php
Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductQueryBuilder.php]: Fixed in the commit [fae4183](https://github.com/woocommerce/woocommerce/pull/52429/commits/fae4183596ec5f4542622906f06f2ce56565b5e2).

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php
Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php]: Method moved in the commit [6f83efb](https://github.com/woocommerce/woocommerce/pull/52429/commits/6f83efb71292e4bea35deadd395f710170e39e32).

**Comment by octaedro** [comment]
Hi @imanish003, thank you for reviewing this PR.
I just addressed the changes you suggested, could you take another look at this PR?

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php
Comment (kmanijak) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php:1]: What about naming the file and class `ProductCollectionRenderer` to keep the convention with `ProductCollectionController`?

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php
Comment (kmanijak) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php]: Considering we have a namespace now, maybe we could include Utils in in as well so there's no dependencies 🤔 On the other hand there's `\Utils` directory specifically for it so not really sure about that. But wanted to raise that option. I kind of feel it would be clearer having all of the Product Collection related files in single place but happy to hear other's opinion.

**Review by kmanijak — COMMENTED** [review]
@octaedro, thank you for working on it! It's really important and valuable PR! I smoke tested a block and can't see any regressions, all seems to be working fine. I left some minor comments about naming and files organisation but overall I'm happy to hear your opinion and approve the PR soon!

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php
Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php]: I think that it's an interesting idea. Here are some pros and cons:

### ✅ Pros:
- Better encapsulation - all Product Collection related code lives together
- Clearer ownership and scope
- Easier to understand what the utils are for
- Makes it harder to accidentally use these utils in unrelated contexts
- Follows Single Responsibility Principle more closely

### ❌ Cons:
- If any utilities are genuinely reusable across different blocks, they become harder to discover
- Goes against the current project structure where utils are centralized
- Might lead to code duplication if similar utilities are needed elsewhere

I lean towards moving the `ProductCollectionUtils` inside `ProductCollection`. However, this PR is already quite large. What do you think about addressing this change in a follow-up PR?

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php
Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php:1]: Good call! Renamed in the commit [f0b7145](https://github.com/woocommerce/woocommerce/pull/52429/commits/f0b714585e0067c3512b405ecffcea70807e93b2)!

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php
Comment (kmanijak) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php]: > However, this PR is already quite large. What do you think about addressing this change in a follow-up PR?

Yeah, I don't think it's a must no matter the final decision!

Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductRenderer.php]: Hi @kmanijak,
I just re-rebased this PR. Could you take another look at it?

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/CollectionHandlerRegistry.php
Comment (imanish003) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/CollectionHandlerRegistry.php:1]: I was searching for this file using "Cmd + P," but I wasn't able to find it because I was looking for the prefix "ProductCollection." Should we consider renaming this file to "ProductCollectionHandlerRegistry.php"(We may also remove "Handler" from name i.e. `ProductCollectionRegistry.php`)? I feel like we will naturally search for the prefix "ProductCollection" to find all files related to Product Collection. What do you think? This change doesn't need to be addressed in this PR itself. We can create a follow-up PR to address this.

**Review by imanish003 — APPROVED** [review]
I smoke tested everything and it works as expected. The separation of concerns and improved code organization make the codebase much more maintainable. Excellent work on this refactor @octaedro! 🚀

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/CollectionHandlerRegistry.php
Comment (kmanijak) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/CollectionHandlerRegistry.php:1]: I had similar thoughts when initially reviewing this PR couple days ago. The only thing that I wasn't sure about was that this is actually Registry of Collection Handlers aka `CollectionHandlerRegistry` 😆 so kind of the name reflects the purpose. Hence I didn't share my thoughts back then.

> I feel like we will naturally search for the prefix "ProductCollection" to find all files related to Product Collection.

True, but at the same time file is in `/ProductCollection` directory and namespace so maybe that's enough to describe it belongs and is used in Product Collection.

To sum up I think I'm fine with keeping the current name but it's not a strong opinion 🙈 

EDIT: Kind of similar to `ProductQueryBuilder` which is also not named `ProductCollectionQueryBuilder` and I think that makes sense 🤔

**Review by kmanijak — APPROVED** [review]
I smoke tested the PR and didn't face any issues so I think it's good to go! Amazing work @octaedro! 🎉

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/CollectionHandlerRegistry.php
Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/CollectionHandlerRegistry.php:1]: > I feel like we will naturally search for the prefix "ProductCollection" to find all files related to Product Collection.

I don't have a strong opinion on this, but it would make sense to follow a single pattern. Since we don't prefix the classes with the parent folder name for [OrderConfirmation](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/src/Blocks/BlockTypes/OrderConfirmation) (a sibling folder of "ProductCollection"), I think it's best to omit the prefix.

In that case, the files would be organized like this:
- ProductCollection
  - HandlerRegistry
  - Controller
  - NoResults
  - Renderer

Does the structure above make sense to you? @imanish003 @kmanijak

Comment (kmanijak) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/CollectionHandlerRegistry.php:1]: I like that approach! The class name becomes quite generic (e.g. `Controller` 😅 ) but at the same time it has a namespace that defines the context to I think it's a perfect way to go!

Let's not forget about ProductQueryBuilder. I think I'd go with Query Builder in that case, so:

- ProductCollection\
    - HandlerRegistry
    - Controller
    - NoResults
    - Renderer
    - QueryBuilder

Comment (imanish003) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/CollectionHandlerRegistry.php:1]: Sounds good to me as well 🙌🏻

Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/CollectionHandlerRegistry.php:1]: > QueryBuilder

Thank you @kmanijak; I forgot to include it in my list 🤦‍♂️

Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/CollectionHandlerRegistry.php:1]: I just renamed the files and classes in the commits [f2c442a](https://github.com/woocommerce/woocommerce/pull/52429/commits/f2c442aebd1e7cf5e1e219241ce45b0528f0c877), [ebd16a4](https://github.com/woocommerce/woocommerce/pull/52429/commits/ebd16a437244c76a6974188253dd3f23ba69682f), [aa3dc35](https://github.com/woocommerce/woocommerce/pull/52429/commits/aa3dc35cb87f792002323a33c3fa181055f9cfe6), [578d69d](https://github.com/woocommerce/woocommerce/pull/52429/commits/578d69d1bcde4fd5fcc94c2892d51ea6f561f144) and [e76ca6c](https://github.com/woocommerce/woocommerce/pull/52429/commits/e76ca6cb97bff566c4f0a619dced0740d636b302).

**Pull request merged**
